File: individual_assigment/code/cpp/matrix_multiplication.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/resource.h>
#include <time.h>
#include <string.h>

typedef struct {
    int size;
    double execution_time;
    double memory_used;
} BenchmarkResult;

double get_memory_usage() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
#ifdef __APPLE__
    return usage.ru_maxrss / (1024.0 * 1024.0);
#else
    return usage.ru_maxrss / 1024.0;
#endif
}

double** create_matrix(int rows, int cols, double value) {
    double** matrix = (double**) malloc(rows * sizeof(double*));
    if (!matrix) return NULL;
    for (int i = 0; i < rows; i++) {
        matrix[i] = (double*) malloc(cols * sizeof(double));
        if (!matrix[i]) return NULL;
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = value;
        }
    }
    return matrix;
}

void free_matrix(double** matrix, int rows) {
    for (int i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

double** matrix_multiply(double** A, double** B, int n, int m, int p) {
    double** C = (double**) malloc(n * sizeof(double*));
    if (!C) return NULL;
    for (int i = 0; i < n; i++) {
        C[i] = (double*) calloc(m, sizeof(double));
        if (!C[i]) return NULL;
        for (int j = 0; j < m; j++) {
            for (int k = 0; k < p; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

BenchmarkResult benchmark_matrix_multiplication(int size) {
    BenchmarkResult result;
    result.size = size;
    double** A = create_matrix(size, size, 1.5);
    double** B = create_matrix(size, size, 2.5);
    double mem_before = get_memory_usage();
    clock_t start = clock();
    double** C = matrix_multiply(A, B, size, size, size);
    clock_t end = clock();
    double mem_after = get_memory_usage();
    result.execution_time = (double)(end - start) / CLOCKS_PER_SEC;
    result.memory_used = mem_after - mem_before;
    free_matrix(A, size);
    free_matrix(B, size);
    free_matrix(C, size);
    return result;
}

void run_benchmarks(int sizes[], int num_sizes) {
    printf("%-10s %-15s %-15s\n", "Size", "Time (s)", "Memory (MB)");
    printf("---------------------------------------------------------------------\n");
    double total_time = 0.0, total_memory = 0.0;
    int successful_runs = 0;
    for (int i = 0; i < num_sizes; i++) {
        int size = sizes[i];
        BenchmarkResult r = benchmark_matrix_multiplication(size);
        total_time += r.execution_time;
        total_memory += r.memory_used;
        successful_runs++;
        printf("%-10d %-15.4f %-15.2f\n", size, r.execution_time, r.memory_used);
    }
    if (successful_runs > 0) {
        printf("\nSUMMARY\n");
        printf("=====================================================================\n");
        printf("Total execution time: %.2f seconds\n", total_time);
        printf("Average memory usage: %.2f MB\n", total_memory / successful_runs);
        printf("Largest matrix tested: %dx%d\n", sizes[successful_runs - 1], sizes[successful_runs - 1]);
    }
}

int main() {
    int test_sizes[] = {64, 128, 256, 512};
    int num_sizes = sizeof(test_sizes) / sizeof(test_sizes[0]);
    run_benchmarks(test_sizes, num_sizes);
    return 0;
}

File: individual_assigment/code/python/matrix_multiplication.py
import time

import tracemalloc
import psutil
import os
from typing import List, Tuple

def matrix_multiply(A: List[List[float]], B: List[List[float]]) -> List[List[float]]:
    """
    Basic O(n³) matrix multiplication algorithm.
    """
    n = len(A)
    m = len(B[0])
    p = len(B)
    
    C = [[0.0 for _ in range(m)] for _ in range(n)]
    
    for i in range(n):
        for j in range(m):
            for k in range(p):
                C[i][j] += A[i][k] * B[k][j]
    
    return C

def create_matrix(rows: int, cols: int, value: float = 1.0) -> List[List[float]]:
    """Create a matrix filled with a specific value."""
    return [[value for _ in range(cols)] for _ in range(rows)]

def benchmark_matrix_multiplication(size: int) -> Tuple[float, float, float]:
    """
    Benchmark matrix multiplication for a given size.
    Returns: (execution_time, memory_used_mb, cpu_percent)
    """
    A = create_matrix(size, size, 1.5)
    B = create_matrix(size, size, 2.5)
    
    process = psutil.Process(os.getpid())
    
    tracemalloc.start()
    
    start_time = time.perf_counter()
    cpu_before = process.cpu_percent(interval=0.1)
    
    C = matrix_multiply(A, B)
    
    end_time = time.perf_counter()
    cpu_after = process.cpu_percent(interval=0.1)
    
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    execution_time = end_time - start_time
    memory_mb = peak / (1024 * 1024)
    avg_cpu = (cpu_before + cpu_after) / 2
    
    return execution_time, memory_mb, avg_cpu

def run_benchmarks(sizes: List[int]):
    """Run benchmarks for multiple matrix sizes."""
    print(f"{'Size':<10} {'Time (s)':<15} {'Memory (MB)':<15} {'CPU %':<10}")
    print("-" * 70)
    
    results = []
    for size in sizes:
        try:
            exec_time, memory, cpu = benchmark_matrix_multiplication(size)
            results.append((size, exec_time, memory, cpu))
            print(f"{size:<10} {exec_time:<15.4f} {memory:<15.2f} {cpu:<10.2f}")
        except MemoryError:
            print(f"{size:<10} MEMORY ERROR - Size too large")
            break
        except Exception as e:
            print(f"{size:<10} ERROR: {str(e)}")
            break
    
    return results

if __name__ == "__main__":
    test_sizes = [64, 128, 256, 512]
    
    
    results = run_benchmarks(test_sizes)
    
    if results:
        print("\n")
        print("SUMMARY")
        print("=" * 70)
        total_time = sum(r[1] for r in results)
        avg_memory = sum(r[2] for r in results) / len(results)
        print(f"Total execution time: {total_time:.2f} seconds")
        print(f"Average memory usage: {avg_memory:.2f} MB")
        print(f"Largest matrix tested: {results[-1][0]}x{results[-1][0]}")
